[JIT] Disable conv-add-relu fusion for cuDNN7 when model uses fp16

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56580 [JIT] Enable conv-add-relu fusion as a part of frozen graph optimization
* **#56579 [JIT] Disable conv-add-relu fusion for cuDNN7 when model uses fp16**

Summary: On earlier cuDNN versions, when a model uses fp16, the performance after conv-add-relu fusion regresses. Let's just disable the fusion for fp16 if cuDNN version is earlier than v8.

Here is the perf number before the change, on Tesla T4, CUDA 10.2, cuDNN 7.6.5. The performance is measured as a speedup of frozen+fusion against frozen, where 1.0x means the same speed and <1.0x means a slowdown after applying fusion.
Benchmark | Speedup against frozen
------------ | -------------
resnet18|0.86x
googlenet|0.86x
alexnet|0.89x
vgg16|0.69x

Test Plan:
Verified for fp16 models on Nvidia Tesla T4 and the speedup is back as 1.0x

Differential Revision: [D27915514](https://our.internmc.facebook.com/intern/diff/D27915514)

desertfire committed:
[JIT] Disable conv-add-relu fusion for earlier cuDNN version

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Disable conv-add-relu fusion for cuDNN7 when model uses fp16"


Summary: On earlier cuDNN versions, when a model uses fp16, the
performance after conv-add-relu fusion regresses. Let's just
disable the fusion for fp16 if cuDNN version is older than v8.

Test Plan:
Tested for fp16 models on Nvidia Tesla T4

[ghstack-poisoned]

Krovatkin reviewed (approved):
:shipit:
Looks reasonable.
Could you please share which models regress and by how much in a comment?

desertfire committed:
Update on "[JIT] Disable conv-add-relu fusion for cuDNN7 when model uses fp16"


Summary: On earlier cuDNN versions, when a model uses fp16, the performance after conv-add-relu fusion regresses. Let's just disable the fusion for fp16 if cuDNN version is earlier than v8.

Here is the perf number before the change, on Tesla T4, CUDA 10.2, cuDNN 7.6.5. The performance is measured as a speedup of frozen+fusion against frozen, where 1.0x means the same speed and <1.0x means a slowdown after applying fusion.
Benchmark | Speedup against frozen
------------ | -------------
resnet18|0.86x
googlenet|0.86x
alexnet|0.89x
vgg16|0.69x

Test Plan:
Verified for fp16 models on Nvidia Tesla T4 and the speedup is back as 1.0x

Differential Revision: [D27915514](https://our.internmc.facebook.com/intern/diff/D27915514)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Disable conv-add-relu fusion for cuDNN7 when model uses fp16"


Summary: On earlier cuDNN versions, when a model uses fp16, the performance after conv-add-relu fusion regresses. Let's just disable the fusion for fp16 if cuDNN version is earlier than v8.

Here is the perf number before the change, on Tesla T4, CUDA 10.2, cuDNN 7.6.5. The performance is measured as a speedup of frozen+fusion against frozen, where 1.0x means the same speed and <1.0x means a slowdown after applying fusion.
Benchmark | Speedup against frozen
------------ | -------------
resnet18|0.86x
googlenet|0.86x
alexnet|0.89x
vgg16|0.69x

Test Plan:
Verified for fp16 models on Nvidia Tesla T4 and the speedup is back as 1.0x

Differential Revision: [D27915514](https://our.internmc.facebook.com/intern/diff/D27915514)

[ghstack-poisoned]

desertfire committed:
Update on "[JIT] Disable conv-add-relu fusion for cuDNN7 when model uses fp16"


Summary: On earlier cuDNN versions, when a model uses fp16, the performance after conv-add-relu fusion regresses. Let's just disable the fusion for fp16 if cuDNN version is earlier than v8.

Here is the perf number before the change, on Tesla T4, CUDA 10.2, cuDNN 7.6.5. The performance is measured as a speedup of frozen+fusion against frozen, where 1.0x means the same speed and <1.0x means a slowdown after applying fusion.
Benchmark | Speedup against frozen
------------ | -------------
resnet18|0.86x
googlenet|0.86x
alexnet|0.89x
vgg16|0.69x

Test Plan:
Verified for fp16 models on Nvidia Tesla T4 and the speedup is back as 1.0x

Differential Revision: [D27915514](https://our.internmc.facebook.com/intern/diff/D27915514)

[ghstack-poisoned]
